tracing-subscriber/src/registry/sharded.rs

```diff
@@ -270,3 +320,100 @@ impl<'a> SpanData<'a> for Data<'a> {
         ExtensionsMut::new(self.inner.extensions.write().expect("Mutex poisoned"))
     }
 }
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::Registry;
+    use crate::{layer::Context, registry::LookupSpan, Layer};
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use tracing::{self, subscriber::with_default};
+    use tracing_core::{
+        span::{Attributes, Id},
+        Subscriber,
+    };
+
+    struct NopLayer;
+    impl<S> Layer<S> for NopLayer
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+            assert!(&ctx.span(&id).is_some());
+        }
+    }
+
+    struct NopLayer2;
+    impl<S> Layer<S> for NopLayer2
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+            assert!(&ctx.span(&id).is_some());
+        }
+    }
+
+    #[test]
+    fn single_layer_can_access_closed_span() {
+        let subscriber = NopLayer
+            .and_then(NopLayer)
+            .with_subscriber(Registry::default());
+
+        with_default(subscriber, || {
+            let span = tracing::debug_span!("span");
+            drop(span);
+        });
+    }
+
+    #[test]
+    fn multiple_layers_can_access_closed_span() {
+        let subscriber = NopLayer
+            .and_then(NopLayer)
+            .and_then(NopLayer2)
+            .with_subscriber(Registry::default());
+
+        with_default(subscriber, || {
+            let span = tracing::debug_span!("span");
+            drop(span);
+        });
+    }
+
+    #[test]
+    fn span_is_removed_from_registry() {
+        static IS_REMOVED: AtomicBool = AtomicBool::new(false);
+
+        struct ClosingLayer;
+        impl<S> Layer<S> for ClosingLayer
+        where
+            S: Subscriber + for<'a> LookupSpan<'a>,
+        {
+            fn new_span(&self, _: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+                let span = ctx.span(id).expect("Missing span; this is a bug");
+                let mut extensions = span.extensions_mut();
+                extensions.insert(ClosingSpan);
+            }
+
+            fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+                assert!(&ctx.span(&id).is_some());
```

Comment:
should we maybe also assert the extension is present here? It should be, though.

Reply:
agreed, we should.

Reply:
Added.